DOC-9214 udfs for query

RayOffiah opened the pull request.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Added page creating-a-javascript-udf.adoc
Renamed images directory

RayOffiah committed:
Added page creating-a-javascript-udf.adoc
Adding diagram for call sequence

RayOffiah committed:
Added page creating-a-javascript-udf.adoc
Adding sql++ caller function

RayOffiah committed:
Added page creating-a-javascript-udf.adoc
Rearranging documents

RayOffiah committed:
Added page creating-a-javascript-udf.adoc
Forgot to change the menu

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Merge branch 'release/7.1' into DOC-9214_UDFs_for_Query

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Merge branch 'release/7.1' into DOC-9214_UDFs_for_Query

RayOffiah committed:
Work in progress

RayOffiah committed:
Changing diagrams to a vertical layout.

RayOffiah committed:
Added error-handling page and a file for further reading

RayOffiah committed:
Fixing the navigation menu

RayOffiah committed:
Finishing off the error-handling section and doing some tidying up.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Adding information about side effects and transactions.
And some tidying up

RayOffiah committed:
Second cut

RayOffiah committed:
Added a note about javascript workers and recursion.

simon-dew commented on modules/guides/pages/javascript-udfs/introduction.adoc:
```diff
@@ -0,0 +1,47 @@
+= User-defined Functions in Javascript
+:description: Extending the {sqlpp} query language by adding your own functions written in Javascript
```
{sqlpp} is not defined

RayOffiah replied:
If you're using a local playbook then you need to define it in there. It's already defined in the playbooks for production.

simon-dew replied:
Hang on a second. This is going to lead to inconsistent usage for some really important terminology. (And for a terminology change that is _still under discussion_.) The attributes can stay here, of course, but those playbooks need to change back to the existing terminology until we have introduced this attribute everywhere it is needed throughout the docs set.

RayOffiah replied:
Mmm. Yeah, having thought about it for a while, this isn't really what attributes should be used for anyway. The big problem is that N1QL, SQL++ or whatever it might be in the future might take an 'a' or an 'an' (and other grammarly things I haven't thought of), so every change should really be checked anyway just to make sure. And if you have to check, then I'm not sure an attribute is going to help that much in this case.

simon-dew replied:
Raised a PR on docs-site.

RayOffiah replied:
👍🏾

simon-dew replied:
> The big problem is that N1QL, SQL++ or whatever it might be in the future might take an 'a' or an 'an'

Yep — I raised this in style and tone. You're right about the proper use of attributes. But if we _don't_ use the attribute, then we would have to make the change all at once, in one intense docs sprint 😬

simon-dew commented on modules/guides/pages/javascript-udfs/introduction.adoc:
```diff
@@ -0,0 +1,47 @@
+* Create reuseable, domain-specific functions for use in your applications.
+* Execute complex logic that may be difficult to do in {sqlpp}.
+* Migrate from RDBMS stored procedures.
+* Use  {sqlpp} statements embedded in your Javascript to access and modify your data.
```
This goes into too much detail for a how-to guide. Should be in a conceptual topic.

simon-dew commented on modules/guides/pages/javascript-udfs/introduction.adoc:
```diff
@@ -0,0 +1,47 @@
+
+== Next Steps
+ 
+'''
```
We don't use horizontal rules to delimit numbered lists anywhere else?

RayOffiah replied:
Yup, that's too many lines at the bottom of the page.

Removing.

simon-dew commented on modules/guides/pages/javascript-udfs/introduction.adoc:
```diff
@@ -0,0 +1,47 @@
+= User-defined Functions in Javascript
```
(1) Title case for headings; (2) "Javascript" should be "JavaScript" throughout.

```suggestion
= User-Defined Functions in JavaScript
```

RayOffiah replied:
Fixed, but I've left it as 'Javascript' where I'm quoting from the UI.

simon-dew commented on modules/guides/partials/javascript-udfs/before-you-begin.adoc:
```diff
@@ -0,0 +1,23 @@
+////
+This is the 'before you begin' section that goes in the introduction.
+We can use it in other places if it's needed.
+////
```
Overall, this is too much explanatory detail for a how-to guide; reads more like a tutorial.

RayOffiah replied:
That's an AsciiDoc comment.

simon-dew replied:
I know, I was referring to the page as a whole.

simon-dew commented on modules/guides/partials/javascript-udfs/before-you-begin.adoc:
```diff
@@ -0,0 +1,23 @@
+We can use it in other places if it's needed.
+////
+
+You're going to need a basic working knowledge of https://www.w3schools.com/js/[Javascript^].
```
Again, reads more like a tutorial. If the user is looking at a how-to guide for creating JavaScript UDFs, we have to assume that they already know about JavaScript. Secondly, I'd prefer to link to the ECMAScript standard or the Wikipedia page for JavaScript than w3schools.

RayOffiah replied:
Removed

simon-dew commented on modules/guides/partials/javascript-udfs/before-you-begin.adoc:
```diff
@@ -0,0 +1,23 @@
+
+You're going to need a basic working knowledge of https://www.w3schools.com/js/[Javascript^].
+
+You're going to be running through some live examples, so will need access to a Couchbase server. 
```
"You're going to be running through some live examples" again, if this is meant to be a how-to guide, they might not be. Reads more like a tutorial.

RayOffiah replied:
Yup, this one's gone.

simon-dew commented on modules/guides/partials/javascript-udfs/further-reading.adoc:
```diff
@@ -0,0 +1,43 @@
+////
+This file is the further reading section at the bottom
+of each page
```
For the other how-to guides we put each related link in a separate bullet point. It's only the links to the SDK languages that are linked within a single bullet point separated by vertical bars.

RayOffiah replied:
Ah, hadn't spotted the distinction. No problem.

simon-dew commented on modules/ROOT/nav.adoc:
```diff
@@ -49,7 +49,12 @@
+    *** xref:guides:javascript-udfs/calling-javascript-udf.adoc[]
+    *** xref:guides:javascript-udfs/calling-sqlpp-from-javascript.adoc[]
+    *** xref:guides:javascript-udfs/data-handling-javascript-udf.adoc[]
+    *** xref:guides:javascript-udfs/handling-errors-javascript-udf.adoc[]
```
Not sure about the inconsistency here. I'd rather have (a) the JavaScript UDF how-to guide pages in the same directory as all the other guides, or (b) separate _all_ the how-to guides into their own directory.

RayOffiah replied:
I'm going to leave this one I think.

It's better to have the guides in separate folders, especially when you have quite a few images and example code running through them It just makes it a lot easier to find the bits you're working on.

I'll have a look at moving the other guides into their own folders once I've got Neo out the way.

simon-dew replied:
I'm going to push back on this a little bit, because IME moving existing files to different locations always leads to broken links and frustrating unforeseen problems. I take your point that it's easier for an author to work on a single isolated set of files. But my (general) observation is that we're never working on a single set of files: they fit into a wider context, with shared images, shared examples, and shared partials, etc. Of course that's harder for the author, because you have to learn that wider context and keep it in your head. But consistency helps. So sorry to be a pain, but I'm going to recommend that you put the UDF guide pages, images, and partials into the same structure as the other existing guides. Examples can or course stay separate, because there are such a lot of them, and the other guide examples are already split up into subdirectories.

Please note the top-level `javascript-udf` module is a great idea and can stay exactly where it is. I'll decide whether it belongs under Develop or Reference later.

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-udf.adoc:
```diff
@@ -0,0 +1,56 @@
+
+== Introduction
+
+We're going to introduce a number of new concepts in this example:
```
How-to guides shouldn't be introducing concepts. That kind of thing should go into a conceptual topic.

simon-dew replied:
Well, to be clear, how-to guides don't shy away from describing concepts completely. Otherwise users would be flipping back and forth between the different pages to understand what they're doing. It's more the implicit thought behind the wording that I'm talking about: the concepts should be _introduced_ in a conceptual topic, and that topic should be linked from the how-to guide. Even if the how-to guide reiterates the concepts in brief.

simon-dew replied:
Note that the "conceptual topic" I keep mentioning should be created as part of DOC-9256.

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-udf.adoc:
```diff
@@ -0,0 +1,56 @@
+[#introduction-create-javascript-library]
+== Creating a Javascript Library
+
+First, you need to create a library to store your Javascript functions.
+
+Select btn:[Query] to access the Query Workbench, then select btn:[UDF] Query Workbench menu.
+
+image::javascript-udfs/navigate-to-udf-query.png[alt="route to the user-defined functions screen"]
+
+You are now in the User-Defined Functions screen. 
+
+image::javascript-udfs/udf-library-screen.png[alt="user-defined functions library screen"]
+
+To create a new function library, click on the *{plus} add function library* link in the `Javascript Function Libraries` table.
+
+The `add library` screens is displayed.
+
+image::javascript-udfs/add-library-screen.png[,600, alt="add screen"]
+
+You have to give the library a name before you can save it; in this example, we've simply called it `my-library`.
+New libraries are created with a simple `add` example function to start with.
+
+You can add your own function, such as the one given below, which will find the number of business days (weekdays) between two dates.
+
+
+[source, javascript]
+----
+include::example$javascript-udfs/get-business-days.js[]
+----
+
+image::javascript-udfs/added-get-business-days-function.png[,600, alt="added new function"]
+
+Now, save the library by pressing the btn:[Save] button.
+
+Your library is created and will show up on the `Javascript Function Libraries`list.
+
+image::javascript-udfs/my-library-list.png[]
```
First thing is, this only covers the UI. In all of the other how-to guides we've covered the UI and the programmatic methods, given that the how-to guides are aimed at developers.

Secondly, need to tighten up the wording to make it into more of a procedure than a tutorial. I've had a stab at rewording it below.

```suggestion
[{tabs}]
====
Query Workbench::
+
--
To create a library to store your JavaScript functions:

1. Select menu:Query[UDF] to display the User-Defined Functions screen.
+
image::javascript-udfs/udf-library-screen.png[alt="User-defined functions library screen"]

// no need to show images of how to select the User-Defined functions screen.
// This should be covered in the Query Workbench documentation by DOC-9596.

2. In the `Javascript Function Libraries` table, click on the *{plus} add function library* link.
+
The `add library` screens is displayed.
New libraries are created with a simple `add` example function to start with.

3. In the *Library Name* box, enter a name for the library.
+
image::javascript-udfs/add-library-screen.png[,600, alt="add screen"]

4. Enter your own JavaScript functions as required.
+ 
image::javascript-udfs/added-get-business-days-function.png[,600, alt="added new function"]

5. When you are ready, save the library by pressing the btn:[Save] button.
+
Your library is created and will show up on the `Javascript Function Libraries`list.
+
image::javascript-udfs/my-library-list.png[]

'''

For example, the function given below finds the number of business days (weekdays) between two dates.

[source, javascript]
----
include::example$javascript-udfs/get-business-days.js[]
----
--
====
```

simon-dew replied:
Note the main difference — in a tutorial, we tell the users what to type. In a how-to guide, we tell them where the commands are, and give them examples if appropriate; but they type what they want.

RayOffiah replied:
All this has been revamped.  Needs reviewing again.

simon-dew commented on modules/guides/pages/javascript-udfs/data-handling-javascript-udf.adoc:
```diff
@@ -0,0 +1,149 @@
+= Data handling in Javascript UDFs
```
Haven't gone into this page in detail but the same comments apply. This is a tutorial rather than a how-to guide.

RayOffiah replied:
Removed page

simon-dew commented on modules/guides/pages/javascript-udfs/handling-errors-javascript-udf.adoc:
```diff
@@ -0,0 +1,70 @@
+= Handling Errors in Javascript UDFs
```
Again, same comments apply: this is a tutorial rather than a how-to guide.

simon-dew commented on modules/guides/examples/javascript-udfs/add-airline-with-date-response.json5:
```diff
@@ -0,0 +1,14 @@
+      "icao": "MLA",
+      "id": 1220,
+      "name": "Raz Air",
+      "opdate": "2022-02-25T09:40:57.589Z",    // <.>
```
I've noticed that you're using the json5 file extension, presumably so that you can include comments in the JSON for asciidoc tagging. But you're not actually using any of the other JSON5 features. Have you considered using jsonc (JSON with comments) as a file extension, for consistency with other example files?

RayOffiah replied:
Yep, can certainly do that.

simon-dew reviewed (changes requested):
Thanks for the work on this Ray. The main thing I would say is that this is a tutorial rather than a how-to guide. I've added some comments to the creating-javascript-udf page to suggest how I would turn it into a how-to guide. I'll send more information on the distinction in a moment.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Revamped the set as a How-To guide.

RayOffiah committed:
Cleaning up redundant files.

RayOffiah committed:
Changing 'Javascript' to 'JavaScript'.

RayOffiah committed:
Work in progress

RayOffiah committed:
DOC-9596 Document UI updates for UDF/JS: Getting started

RayOffiah committed:
Work in progress

RayOffiah committed:
Added the 'remove' stuff, and some extras around the CBQ shell.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Revert "Work in progress"

This reverts commit 123427266272533ddb88d73d7b6ca0d9955fc156.

RayOffiah committed:
Revert "Revert "Work in progress""

This reverts commit 46961338f3e3621d981f204e09b6ff7cf99f4737.

RayOffiah committed:
Reverting the cbq shell anchor addition.

RayOffiah committed:
Reorganising the pages

RayOffiah committed:
Sorting out some scoping stuff.

RayOffiah committed:
Adding scoped function removal

RayOffiah committed:
Fixing diagram in diagrams-plantuml.adoc

RayOffiah committed:
DOC-9596 Document UI updates for UDF/JS

RayOffiah committed:
Added page on handling data in Javascript

RayOffiah committed:
Callout list was in the wrong place.

RayOffiah committed:
Callout list was in the wrong place.

RayOffiah committed:
Fixing inclusion problem.

RayOffiah committed:
Added Javascript-JSON basics section

RayOffiah committed:
DOC-9352: Getting the name of the Capella instance

RayOffiah committed:
Fixing the menu

RayOffiah committed:
DOC-9466 [N1QL-UDF] if a UDF statement calls another UDF statement, there is a dynamic limit on t...
Added recursion limit entry along and separated the eventing stuff from the basic UDF stuff.

RayOffiah committed:
WIP

RayOffiah committed:
Merge branch 'release/7.1' into DOC-9214_UDFs_for_Query

RayOffiah committed:
Fixes from review comments.

simon-dew commented on modules/guides/examples/javascript-udfs/create-javascript-function-with-rest.sh:
```diff
@@ -0,0 +1,14 @@
+curl -v POST http://localhost:8093/evaluator/v1/libraries/my-library  \   # <.>
+ -u Administrator:password \    # <.>
+ -d 'function getBusinessDays(startDate, endDate) {    # <.>
```
Just checking that you've tested this script — does the bash comment at the end of this line cause problems inside the JavaScript payload? (Yes, the comment gets converted into a callout by AsciiDoc, but that's not the issue: when automated testing runs this script, the bash comment is going to get included with the JavaScript payload.)

simon-dew commented on modules/guides/examples/javascript-udfs/create-reference-function.n1sql:
```diff
@@ -0,0 +1,2 @@
+CREATE FUNCTION GetBusinessDays(startDate, endDate) LANGUAGE JAVASCRIPT AS "getBusinessDays" AT "my-library;
```
File extension should be `.n1ql`

simon-dew commented on modules/guides/examples/javascript-udfs/create-reference-function.sh:
```diff
@@ -0,0 +1,3 @@
+curl -v http://localhost:8093/query/service \
+  -u Administrator:password \
+  -d 'statement=CREATE FUNCTION GetBusinessDays(startDate, endDate) LANGUAGE JAVASCRIPT AS "getBusinessDays" AT "my-library"'
```
Best practice: finish all N1QL statements with a semicolon

```suggestion
  -d 'statement=CREATE FUNCTION GetBusinessDays(startDate, endDate) LANGUAGE JAVASCRIPT AS "getBusinessDays" AT "my-library";'
```

simon-dew commented on modules/guides/examples/javascript-udfs/create-scoped-sqlpp-reference-function-with-rest.sh:
```diff
@@ -0,0 +1,4 @@
+curl -v http://localhost:8093/query/service \
+  -u Administrator:password \
+  -d 'statement=CREATE FUNCTION default:`travel-sample`.inventory.GetBusinessDays(...) #<.>
```
Bash comment inside the N1QL payload. That is going to cause problems for N1QL when automated testing tries to run this script.

simon-dew commented on modules/guides/examples/javascript-udfs/drop-reference-function.n1sql:
```diff
@@ -0,0 +1,2 @@
+DROP FUNCTION GetBusinessDays;
+
```
File extension should be `.n1ql`

simon-dew commented on modules/guides/examples/javascript-udfs/drop-sqlpp-reference-function.sh:
```diff
@@ -0,0 +1,3 @@
+curl -v http://localhost:8093/query/service \
+  -u Administrator:password \
+  -d 'statement=DROP FUNCTION GetBusinessDays'
```
Best practice: semicolon at end of N1QL statement

```suggestion
  -d 'statement=DROP FUNCTION GetBusinessDays;'
```

simon-dew commented on modules/guides/examples/javascript-udfs/execute-get-business-days.n1ql:
```diff
@@ -0,0 +1 @@
+SELECT GetBusinessDays('02/14/2022', '4/16/2022')
```
Best practice: semicolon at end of N1QL statement

```suggestion
SELECT GetBusinessDays('02/14/2022', '4/16/2022');
```

simon-dew commented on modules/guides/examples/javascript-udfs/execute-javascript-function.sh:
```diff
@@ -0,0 +1,3 @@
+curl -v http://localhost:8093/query/service \
+  -u Administrator:password \
+  -d 'statement=EXECUTE FUNCTION GetBusinessDays("02/14/2022", "04/16/2022")'
```
Best practice: semicolon at end of N1QL statement

```suggestion
  -d 'statement=EXECUTE FUNCTION GetBusinessDays("02/14/2022", "04/16/2022");'
```

simon-dew commented on modules/guides/examples/javascript-udfs/execute-scoped-function.n1ql:
```diff
@@ -0,0 +1 @@
+EXECUTE FUNCTION default:`travel-sample`.`inventory`.GetBusinessDays("03/10/2022", "05/10.2022");
```
Does `inventory` need to be in backticks here?

simon-dew commented on modules/guides/examples/javascript-udfs/execute-scoped-function.sh:
```diff
@@ -0,0 +1,3 @@
+curl -v http://localhost:8093/query/service \
+  -u Administrator:password \
+  -d 'statement=EXECUTE FUNCTION default:`travel-sample`.`inventory`.GetBusinessDays("03/10/2022", "05/10.2022")'
```
Best practice: semicolon at end of N1QL statement

```suggestion
  -d 'statement=EXECUTE FUNCTION default:`travel-sample`.`inventory`.GetBusinessDays("03/10/2022", "05/10.2022");'
```

RayOffiah committed:
DOC-9861 JS-UDFs : Document limitations on language constructs

simon-dew commented on modules/guides/examples/javascript-udfs/select-true-alias-get-business-days.n1ql:
```diff
@@ -0,0 +1,5 @@
+  WHEN  GetBusinessDays('02/14/2022', '4/16/2022') > 44 THEN "true" 
+  ELSE "false" 
+  END 
+  AS response    #<.>
```
Semicolon at end of statement. Use N1QL in-line comment syntax

```suggestion
  AS response;    -- <.>
```

simon-dew commented on modules/guides/examples/javascript-udfs/select-true-get-business-days.n1ql:
```diff
@@ -0,0 +1 @@
+SELECT "true" where GetBusinessDays('02/14/2022', '4/16/2022') > 44
```
Coding style for examples: semicolon at end of N1QL statement

```suggestion
SELECT "true" where GetBusinessDays('02/14/2022', '4/16/2022') > 44;
```

simon-dew commented on modules/guides/pages/get-capella-host-name.adoc:
```diff
@@ -0,0 +1,29 @@
+= Getting the Capella Host Name
```
Remove this file and its associated images from this PR. But stash them for later: we'll need them for the Capella docs.

simon-dew commented on modules/guides/pages/intro.adoc:
```diff
@@ -61,4 +61,10 @@ This guide explains how to carry out ACID transactions on the database.
+
+This guide shows explains how to retrieve the host name of a Capella Cloud instance so that you can execute API calls against the cloud service instance.
+
+* xref:get-capella-host-name.adoc[]
```
Revert this change.

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-function-within-scope.adoc:
```diff
@@ -0,0 +1,62 @@
+= Calling a Function Defined within a Namespace
```
Combine this file with `calling-javascript-udf.adoc`. It's confusing for the user to have two how-to guides covering almost identical procedures.

RayOffiah committed:
Changing SQL++ references to N1QL

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+--
+. Click on menu:Workbench[] to access the Query Editor.
++
+image::javascript-udfs/workbench-menu-item.png[alt="select Workbench menu item"]
```
Delete this. For a how-to guide, we have to assume that they already know how to access the Query Workbench. Otherwise every guide would be as long as your arm. If necessary, users can access this information from the "Before You Begin" section at the start of each how-to guide.

RayOffiah committed:
And attribute settings in diagrams are no longer needed.

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
++
+image::javascript-udfs/workbench-menu-item.png[alt="select Workbench menu item"]
+
+. Enter an {sqlpp} statement to call your function:
```
I think we have to assume that SQL++ is pronounced "sequel plus plus", rather than "es queue ell plus plus". Otherwise we're getting into the realms where transclusion meets morphosyntax, which AsciiDoc is really _not_ set up to cope with 😉 

```suggestion
. Enter a {sqlpp} statement to call your function:
```

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+
+. Enter an {sqlpp} statement to call your function:
++
+image::javascript-udfs/add-statement-in-editor.png[ ,600, alt="adding statement to query editor"]
```
We always use a code listing for the Query Workbench examples, rather than an image, so that users can copy and paste the example straight in. The example file should be the same file as the one used for the cbq example. If necessary, use tags to exclude any cbq commands.

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+
+. Set the `context` to `unset` so that we can run the query under the global context.
++
+image::javascript-udfs/setting-context.png[alt="setting the context to unset"]
```
We have a common image file and standard wording for this.

RayOffiah replied:
Where is that exactly?

I've found https://docs.couchbase.com/server/current/tools/query-workbench.html#query-context

Is that it?

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+
+Now press the btn:[Execute] button to run the statement.
+
+The statement will, making a call to the JavaScript function and return a result:
```
```suggestion
The statement makes a call to the JavaScript function and returns a result:
```

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+----
+include::example$javascript-udfs/execute-javascript-function.sh[]
+----
+include::partial$javascript-udfs/execute-sqlpp-result-explanation.adoc[]
```
We wouldn't normally go into the explanations of the metrics block here: this is covered elsewhere.

simon-dew commented on modules/guides/pages/javascript-udfs/create-javascript-function-within-scope.adoc:
```diff
@@ -0,0 +1,125 @@
+= Creating a JavaScript UDF Within a Scope
```
Combine this file with `creating-javascript-library.adoc`. It's confusing for the user to have two how-to guides covering almost identical procedures.

simon-dew commented on modules/guides/pages/javascript-udfs/handling-data-in-javascript-udf.adoc:
```diff
@@ -0,0 +1,21 @@
+= Data handling in Javascript UDFs
```
This section contains important concepts, but I don't think they belong in the how-to guides. I'll make suggestions on the other pages in this section for where to move them. I would delete this page, because outside the context of the how-to guides it's not actually needed.

simon-dew commented on modules/guides/pages/javascript-udfs/javascript-and-json-objects.adoc:
```diff
@@ -0,0 +1,89 @@
+= JavaScript and JSON Objects
```
Can you please move this page to the top-level `javascript-udfs` module for now? I'll decide later whether that module should be under Reference or Develop.

simon-dew commented on modules/guides/pages/javascript-udfs/using-sqlpp-inside-javascript-udfs.adoc:
```diff
@@ -0,0 +1,149 @@
+= Using N1QL inside JavaScript UDFs
```
Can you please move this page to the top-level `javascript-udfs` module for now? I'll decide later whether that module should be under Reference or Develop.

simon-dew commented on modules/guides/pages/javascript-udfs/using-javascript-library-hierarchies.adoc:
```diff
@@ -0,0 +1,28 @@
+= Using JavaScript Library Hierarchies
```
This section describes an important concept, but I think it's confusing having a separate, parallel set of how-to guides to cover it. I've suggested combining the guides in this section with the global guides for creating, calling, and removing UDFs.

Can you please rework this page as a concept doc (with links to the how-to guides where necessary) and move it to the top-level `javascript-udfs` module for now? I'll decide later whether that module should be under Reference or Develop.

simon-dew commented on modules/guides/partials/javascript-udfs/create-function-breakdown.adoc:
```diff
@@ -0,0 +1,16 @@
+make up the CREATE function example.
+////
+
+The `CREATE FUNCTION` in this example can be broken down as follows: +
+{nbsp}
+[horizontal]
+`GetBusinessDays(startDate,{nbsp}endDate)`:: This is the name of the N1QL reference function that will be called in your N1QL statements.
+
+`getBusinessDays`:: This is the name of the JavaScript function.
+
+`my-library`:: The name of the library where the JavaScript function resides.
```
Could this be a callout list?

RayOffiah replied:
No, it's used as a CURL command, so it's best to leave it as it is. 
Must have a chat with Dan about this at some point.

RayOffiah replied:
Not really. It is used by a CURL command.

simon-dew commented on modules/guides/partials/javascript-udfs/create-function-breakdown.adoc:
```diff
@@ -0,0 +1,16 @@
+
+`my-library`:: The name of the library where the JavaScript function resides.
+
+For more information on the `CREATE FUNCTION` statement, see xref:n1ql:n1ql-language-reference/createfunction.adoc[]
```
Full stop at end

```suggestion
For more information on the `CREATE FUNCTION` statement, see xref:n1ql:n1ql-language-reference/createfunction.adoc[].
```

simon-dew commented on modules/javascript-udfs/examples/add-airline-with-side-effect.n1ql:
```diff
@@ -0,0 +1 @@
+SELECT "true" as response where AddAirline() = "missing"
```
Coding style for N1QL examples: semicolon at end. Also I'd prefer capitalization for N1QL keywords to make them stand out from identifiers and aliases.

```suggestion
SELECT "true" AS response WHERE AddAirline() = "missing";
```

simon-dew commented on modules/javascript-udfs/examples/execute-do-recursion.n1ql:
```diff
@@ -0,0 +1 @@
+EXECUTE FUNCTION doRecursion()
```
Coding style for N1QL examples: semicolon at end

```suggestion
EXECUTE FUNCTION doRecursion();
```

simon-dew commented on modules/javascript-udfs/pages/built-in-functions.adoc:
```diff
@@ -0,0 +1,152 @@
+
+NOTE: In addition the _N1qlQuery()_ is now deprecated and has been replaced with the _N1QL()_ call which has a different parameter format.
+
+{nbsp} 
```
We should fix the CSS for horizontal definition lists 😉

simon-dew commented on modules/javascript-udfs/pages/built-in-functions.adoc:
```diff
@@ -0,0 +1,152 @@
++
+NOTE: adding an optimal index to the `travel-sample`.`_default.`_default` keyspace for the below query can increase the performance by 57X.
++
+_iterator using a positional params array_
```
I guess you've copied this from the Eventing pages, but is this meant to be a title for the following code sample? Better to use AsciiDoc title markup

simon-dew commented on modules/javascript-udfs/pages/built-in-functions.adoc:
```diff
@@ -0,0 +1,152 @@
+    );
+----
++
+_Example iterator using a named params object_
```
Likewise, is this meant to be a title for the following code sample? Better to use AsciiDoc title markup

simon-dew commented on modules/javascript-udfs/pages/built-in-functions.adoc:
```diff
@@ -0,0 +1,152 @@
+    );
+----
+
+* _options_
```
The switch from definition list to bulleted list is confusing, I'm not sure what it means. Can we rework it as a nested definition list? Would also lead to much cleaner markup.

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+
+== Introduction
+
+As well as being able to call Javascript functions from N1QL, you can also call N1QL statements from inside your Javascript functions.
```
Capitalize JavaScript correctly throughout and replace N1QL with `{sqlpp}` throughout, unless of course you're referring to the function call.

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+
+As shown in the <<calling-statements-inline, examples above>>, embedded N1QL statements  return values which can be used later on in your code.
+
+The values returned from the statement calls are Javascript https://www.w3schools.com/js/js_object_iterables.asp[iterators^]: lists of values or documents returned from the database. 
```
Is there a more definitive link than the w3schools description?

RayOffiah replied:
The w3schools one usually comes near the top of the list, probably because you can try out code.

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+== Passing Parameters to N1QL statements
+
+You can pass parameters from your Javascript to your N1QL statements.
+Parameters can either be `positional` or `named`.
```
Italics for introducing / defining terms. Can use `[.def]` role if you want but it's not recommended.

```suggestion
Parameters can either be _positional_ or _named_.
```

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+[NOTE]
+====
+N1QL calls support both `named` and `positional` parameters.
+Inline calls only support `named` parameters.
```
```suggestion
N1QL calls support both _named_ and _positional_ parameters.
Inline calls only support named parameters.
```

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+include::partial$diagrams.adoc[tags="transactions-and-iterators"]
+====
+
+== Role-base Access Control
```
```suggestion
== Role-Based Access Control
```

simon-dew commented on modules/javascript-udfs/pages/calling-sqlpp-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+====
+The JavaScript workers are created when the Couchbase server is started up.
+
+latexmath:[Number\:of\:JavaScript\:workers = 4 \times number\:of\:CPUs]
```
Nice 👍

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,398 @@
+    var strong = 70;
+    var results =
+        SELECT *                               /* N1QL queries are embedded directly.    */
+        FROM `beer-sample`._default._default   /* Token escaping is standard N1QL style. */
```
We're trying to move away from using `beer-sample` in examples

Don't worry about this too much for now.

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,398 @@
+[source, N1QL]
+----
+SELECT * FROM beersample._default._default WHERE type ...
+----
```
We're trying to move away from using `beer-sample` in examples

Don't worry about this too much for now.

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,398 @@
+|
+|===
+
+*What Happens If You Use a Reserved Word?*
```
Use a discrete heading rather than bolding a paragraph. (I know this probably came from the Eventing pages.)

```suggestion
[discrete]
===== What Happens If You Use a Reserved Word?
```

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,398 @@
+----
+
+During the Function deployment step, when the system validates the function's JavaScript code, it displays an error message such as the following:
++
```
This `+` comes out in the output. Replace with blank line.

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,398 @@
+
+Reserved words as a property bindings value
+
+image::reserved-words-7_0.png[,642]
```
Don't know what's going on with this image. I assume the red text is not in the original. Replace with screenshot with callouts?

RayOffiah replied:
Yes, the red text is in the original, and is correct.

They were demonstrating what the UI does if you use GRANT as an identifier as it happens to be a reserved word.

simon-dew commented on modules/ROOT/nav.adoc:
```diff
@@ -49,7 +49,18 @@
+      **** xref:guides:javascript-udfs/create-javascript-function-within-scope.adoc[]
+      **** xref:guides:javascript-udfs/calling-javascript-function-within-scope.adoc[]
+      **** xref:guides:javascript-udfs/remove-scoped-javascript-function.adoc[]
+  ** xref:guides:get-capella-host-name.adoc[]
```
See comments throughout for changes that will affect the nav.

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+A JavaScript function can be called like any other N1QL function using a suitable N1QL statement.
+The Javascript function is not called directly; it is called through a N1QL reference function.
+
+See xref:javascript-udfs/creating-javascript-library.adoc[]
```
We wouldn't normally add a link to an adjacent how-to guide — the pagination links already do this.

simon-dew commented on modules/guides/pages/javascript-udfs/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,76 @@
+--
+====
+
+
```
Add a link to [User-Defined Functions](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/userfun.html) reference page

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+
+Javascript libraries can be created using the Query Workbench UI or the REST API.
+
+NOTE: You cannot create JavaScript libraries using the CBQ shell.
```
It's more accurate to say you cannot create JavaScript libraries using N1QL.

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
++
+image::javascript-udfs/navigate-to-udf-query.png[alt="route to the user-defined functions screen"]
+
+. Click on the *{plus} add function library* link in the `Javascript Function Libraries` table to show the `Add Library` screen.
```
Bold for UI elements

```suggestion
. Click on the *{plus} add function library* link in the *Javascript Function Libraries* table to show the *Add Library* screen.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+. Click on the *{plus} add function library* link in the `Javascript Function Libraries` table to show the `Add Library` screen.
+
+. Ensure the `Namespace` is set to `global`.
+. Enter a name for the library in the `Library Name` field.
```
Bold for UI elements. (Also reword this to incorporate scoped libraries.)

```suggestion
. Ensure the *Namespace* is set to `global`.
. Enter a name for the library in the *Library Name* field.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+
+. Add your own function to the library by typing it in below the preset example.
++
+image::javascript-udfs/added-get-business-days-function.png[,600, alt="added new function"]
```
We always use a code listing for the UI examples, rather than an image, so that users can copy and paste the example straight in. The example file, by necessity, does not need to be the same file as the one used for the REST API example.

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+image::javascript-udfs/added-get-business-days-function.png[,600, alt="added new function"]
+
+. Save the library by pressing the btn:[Save] button.
+--
```
Include a link to your documentation for the UDF UI!

RayOffiah replied:
I'm going to have to merge the files in from another branch for that.

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+<.> Supply the username and password for your server.
+<.> Supply the function for adding into the library.
+--
+====
```
Include a link to the Functions REST API documentation!

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+== Create the Reference Function
+
+To call a JavaScript function, you first have to add a N1QL function to reference it.
+This can be done through the UI, the REST API or the CBQ Shell.
```
Again, more accurate to say:

```suggestion
This can be done through the UI, the REST API, or a N1QL query.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+Query Workbench:: 
++
+--
+. Click on the `{plus}{nbsp}add function` link from the `UDFs` screen.
```
Bold for UI elements

```suggestion
. Click on the *{plus}{nbsp}add function* link from the *UDFs* screen.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
++
+image::javascript-udfs/my-library-list-add-function-button.png[alt="library list"]
++
+which will display the `Add Function` screen.
```
```suggestion
The *Add Function* dialog is displayed.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+
+*Function Type*:: Set this to `Javascript`. 
+
+*Javascript Library*:: This of name of the JavaScript library where your function is stored.
```
```suggestion
*Javascript Library*:: The name of the JavaScript library where your function is stored.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+
+*Javascript Library*:: This of name of the JavaScript library where your function is stored.
+
+*Library Function Name*:: The name of the JavaScript function in `my-library` that this definition will call.
```
```suggestion
*Library Function Name*:: The name of the JavaScript function in the specified library that this definition will call.
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+=====
+--
+
+CBQ Shell::
```
```suggestion
{sqlpp}::
```

simon-dew commented on modules/guides/pages/javascript-udfs/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,140 @@
+include::partial$javascript-udfs/start-cbq-shell-instructions.adoc[]
+. Enter the N1QL statement to create the reference function:
++
+[source, console]
```
```suggestion
[source, n1ql]
```

simon-dew commented on modules/guides/pages/javascript-udfs/remove-javascript-udf.adoc:
```diff
@@ -0,0 +1,73 @@
+image::javascript-udfs/removing-javascript-function.png[,600]
+
+. Delete your function from the library.
+. Press the btn:[Save] button.
```
Actually there is a REST API call for this which we should cover. See the documentation for [Create or Update a Library](https://docs.couchbase.com/server/current/n1ql/n1ql-rest-api/functions.html#_post_library).

simon-dew commented on modules/guides/pages/javascript-udfs/remove-javascript-udf.adoc:
```diff
@@ -0,0 +1,73 @@
+. Press the btn:[OK] button to confirm that you wish to delete the reference function.
+--
+
+REST API::
++
+--
+Start a command shell and execute the following command:
+
+[source, console]
+----
+include::example$javascript-udfs/drop-sqlpp-reference-function.sh[]
+----
```
And conversely, we don't need an explicit section for this here. It's just the REST API call to execute a query, which is the same as the following section.

simon-dew commented on modules/guides/pages/javascript-udfs/remove-javascript-udf.adoc:
```diff
@@ -0,0 +1,73 @@
+----
+--
+
+CBQ Shell::
```
```suggestion
{sqlpp}::
```

simon-dew commented on modules/guides/pages/javascript-udfs/remove-scoped-javascript-function.adoc:
```diff
@@ -0,0 +1,74 @@
+= Removing a Namespaced JavaScript Function
```
Combine this file with `remove-javascript-udf.adoc`. It's confusing for the user to have two how-to guides covering almost identical procedures.

RayOffiah replied:
Done

simon-dew commented on modules/guides/partials/javascript-udfs/start-cbq-shell-instructions.adoc:
```diff
@@ -0,0 +1,6 @@
+////
```
You don't need this partial every single time the users need to access cbq. Please take a look at the way the other how-to guides handle this.

simon-dew commented on modules/guides/partials/javascript-udfs/what-you-need.adoc:
```diff
@@ -0,0 +1,28 @@
+////
```
Still way to much information in this page for each how-to guide. Why not use the same partial that all the other how-to guides use? (`before-you-begin.adoc`)

simon-dew reviewed (changes requested):
Sorry for the size of this Ray. The core tasks are looking _much_ better as how-to guides — fantastic. We also have a good start on the reference / concept material. I have suggested combining some how-to guides, and moving some information from the guides to the reference / concept section.

I do feel that we're going to need a Learn topic to link all of this stuff together. Thanks for your patience on this.

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tools/pages/udfs-ui.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Added UDF note to the tools-ref.adoc

RayOffiah committed:
Apply suggestions from code review

Committing suggestions.

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Doing the basic fixes

RayOffiah committed:
Moved guide into parent folder

RayOffiah committed:
Changes for review.

RayOffiah committed:
Removing the scoped guides; just incorporate them into the regular ones.

RayOffiah committed:
Adding links and making corrections.

RayOffiah committed:
Merge branch 'DOC-9596_Document_UI_updates_for_UDF_JS' into DOC-9214_UDFs_for_Query

RayOffiah committed:
Adding section about removing Javascript library from server.

RayOffiah committed:
Moving some of the eventing details back to the events section, but linking from the general Javascript UDF page

simon-dew commented on modules/guides/pages/creating-javascript-library.adoc:
```diff
@@ -0,0 +1,130 @@
+----
+--
+
+CBQ Shell::
```
It could be CBQ or in the Query Workbench

```suggestion
N1QL::
```

simon-dew commented on modules/guides/pages/calling-javascript-udf.adoc:
```diff
@@ -0,0 +1,74 @@
+----
+--
+
+CBQ::
```
It could be CBQ or the Query Workbench &mdash; or an SDK call?

```suggestion
N1QL::
```

simon-dew commented on modules/guides/pages/remove-javascript-udf.adoc:
```diff
@@ -0,0 +1,112 @@
+= Removing a JavaScript UDF
+:description: Removing a JavaScript UDF using either the UI, the REST API or the CBQ Shell.
+:page-pagination: 
```
It's the last guide in this section

```suggestion
:page-pagination: prev
```

simon-dew commented on modules/guides/pages/remove-javascript-udf.adoc:
```diff
@@ -0,0 +1,112 @@
+----
+--
+
+CQB Shell::
```
It might be CBQ shell, Query Workbench, N1QL REST API call, SDK...

```suggestion
N1QL::
```

simon-dew commented on modules/javascript-udfs/pages/built-in-functions.adoc:
```diff
@@ -0,0 +1,136 @@
+= Built-In Functions
+:description: A library of JavaScript functions for use with the CouchBase Server
+:page-pagination:
```
We don't tend to use pagination in the develop or reference pages, as the assumption is that users may land on any page as the result of search. At the very least, however, this page should not have the "next" link, as it's the last page in the section.

```suggestion
:page-pagination: prev
```

simon-dew commented on modules/javascript-udfs/pages/calling-n1ql-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+As well as being able to call JavaScript functions from N1QL, you can also call N1QL statements from inside your Javascript functions.
+
+[#calling-statements-inline]
+== Calling N1QL statements inline
```
Just check the capitalization of headings throughout this page (sentence case → title case).

simon-dew commented on modules/javascript-udfs/pages/javascript-and-json-objects.adoc:
```diff
@@ -0,0 +1,89 @@
+
+Couchbase uses JSON has its native data format, so when building JavaScript User-Defined functions, it's important to know how to build JSON data structures natively in JavaScript.
+
+You can read more about the JSON data format at https://www.json.org/[json.org^]
```
Full stop at end

```suggestion
You can read more about the JSON data format at https://www.json.org/[json.org^].
```

simon-dew commented on modules/javascript-udfs/pages/calling-n1ql-from-javascript.adoc:
```diff
@@ -0,0 +1,180 @@
+= Calling N1QL from Javascript
```
Check capitalization of JavaScript throughout this page

```suggestion
= Calling N1QL from JavaScript
```

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,258 @@
+Using JavaScript, you can write your custom Functions.
+Couchbase Functions inherit support for most ECMAScript constructs by using Google v8 as the execution container.
+However, to support the ability to shard and scale Function-execution automatically, some capabilities have been removed.
+Additionally, to optimize language-utilization of the server environment, some new constructs have been added:
```
Not sure if this should be a full stop

```suggestion
Additionally, to optimize language-utilization of the server environment, some new constructs have been added.
```

simon-dew commented on modules/javascript-udfs/pages/javascript-udf-language-constructs.adoc:
```diff
@@ -0,0 +1,258 @@
+*  Basic Keyspace Accessors
+*  Advanced Keyspace Accessors
+
+Details on xref:eventing:eventing-language-constructs.adoc#bucket_accessors[Basic Keyspace Accessors] and xref:eventing:eventing-language-constructs.adoc#advanced_bucket_accessors[Advanced Keyspace Accessors] can be found in xref:eventing:eventing-language-constructs.adoc[]
```
Needs full stop

```suggestion
Details on xref:eventing:eventing-language-constructs.adoc#bucket_accessors[Basic Keyspace Accessors] and xref:eventing:eventing-language-constructs.adoc#advanced_bucket_accessors[Advanced Keyspace Accessors] can be found in xref:eventing:eventing-language-constructs.adoc[].
```

simon-dew commented on modules/javascript-udfs/pages/using-javascript-library-hierarchies.adoc:
```diff
@@ -0,0 +1,23 @@
+= Using JavaScript Library Hierarchies
```
Is this page still meant to exist?

simon-dew commented on modules/javascript-udfs/pages/using-n1ql-inside-javascript-udfs.adoc:
```diff
@@ -0,0 +1,145 @@
+= Using N1QL inside JavaScript UDFs
```
I'd be tempted to change this to "embedding" &mdash; I wasn't sure what the difference was between this page and "Calling N1QL from JavaScript", but the word "embedding" in the abstract made it clear

```suggestion
= Embedding N1QL in JavaScript UDFs
```

simon-dew commented on modules/javascript-udfs/pages/using-n1ql-inside-javascript-udfs.adoc:
```diff
@@ -0,0 +1,145 @@
+
+When using the function as part of a statement call, the unpacking processing is handled by the query engine, so that a query such as:
+
+[source, N1SQL]
```
Should be N1QL

```suggestion
[source, N1QL]
```

simon-dew reviewed (changes requested):
Just a final round of minor corrections. When these are done I think the UDF UI page and the how-to guides are good to go 👍 

I'm also happy with the content of the JavaScript UDF reference, but I understand Binh and Marco still may have some feedback on those pages? I'll decide whether to move the JavaScript UDF reference to the Reference section when you've finished working on it.

I may also make some minor changes to harmonize with the existing documentation after merge, if necessary.

RayOffiah committed:
Apply suggestions from code review

Committed suggestions before next round of changes

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Restoring eventing-language-constructs.adoc

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Moving JS-UDF section under the Query

RayOffiah committed:
Refactoring and adding page to cover calling the UDF (with variadic parameters)

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Fixing menu

RayOffiah committed:
Added section on using scopes and security; a few notes on variadic parameters.

RayOffiah committed:
Fixing error in link

RayOffiah committed:
Renamed 'Reference function' to 'N1QL User-Defined Function'

RayOffiah committed:
Added a 'language constructs' section.

RayOffiah committed:
Tidy up from review number 2

RayOffiah committed:
Added global variables to unsupported constructs

RayOffiah committed:
Added note about not supporting arrow functions.

RayOffiah committed:
Removed javascript-and-json-objects.adoc

RayOffiah committed:
Fixing link

RayOffiah committed:
Added note about logging being unsupported

RayOffiah committed:
Separating the creation of the javascript library from the creation of the UDF.

RayOffiah committed:
Removing remove-javascript-udf.adoc.
Adding callouts

RayOffiah committed:
Incorporating Binh Le's comments.

RayOffiah committed:
Finished review changes

RayOffiah committed:
Changes from second review.

RayOffiah committed:
Changes from second review.

RayOffiah committed:
Quick fix

RayOffiah committed:
Fixes following Marco's email:

-----
"Along with global state, arrow functions are also not supported."

should read

Along with global state global arrow functions are not supported. Arrow functions local to individual javascript functions are supported.
-----

RayOffiah committed:
Fixes from Marco's review (part 1)

RayOffiah committed:
Fixes from Marco's review (part 2)

RayOffiah committed:
Tidying up
